### PR TITLE
Add `Iterator::for_each`

### PR DESCRIPTION
**Pull request by cuviper**
This works like a `for` loop in functional style, applying a closure to
every item in the `Iterator`.  It doesn't allow `break`/`continue` like
a `for` loop, nor any other control flow outside the closure, but it may
be a more legible style for tying up the end of a long iterator chain.

This was tried before in #14911, but nobody made the case for using it
with longer iterators.  There was also `Iterator::advance` at that time
which was more capable than `for_each`, but that no longer exists.

The `itertools` crate has `Itertools::foreach` with the same behavior,
but thankfully the names won't collide.  The `rayon` crate also has a
`ParallelIterator::for_each` where simple `for` loops aren't possible.

> I really wish we had `for_each` on seq iterators. Having to use a
> dummy operation is annoying.  - [@nikomatsakis][1]

[1]: https://github.com/nikomatsakis/rayon/pull/367#issuecomment-308455185

### CODE REVIEW

**Comment by cuviper** [comment]
@bluss I'm curious about your ["interesting reasons"](https://github.com/bluss/rust-itertools/commit/fb513aa7f6e097bd1f1ac763678271b6a14ac5ec) to use `fold` in `Itertools::foreach`.  What sort of optimizations come out of that?

**Comment by alexcrichton** [comment]
👍 I'm game!

@rust-lang/libs, any others have thoughts?

**Comment by aturon** [comment]
Works for me!

**Comment by sfackler** [comment]
I seem to remember there being philosophical objections to this back in the day, but I don't feel particularly strongly.

**Comment by scottmcm** [comment]
@cuviper I think this is the context: https://medium.com/@veedrac/rust-is-slow-and-i-am-the-cure-32facc0fdcb (also https://internals.rust-lang.org/t/pre-rfc-fold-ok-is-composable-internal-iteration/4434)

**Comment by cuviper** [comment]
Interesting - I'll have to play around with that. And if we do want this
implemented on fold, that needs to happen before stabilization.

File: src/libcore/iter/iterator.rs [hunk 1]
```diff
@@ -482,6 +482,52 @@ pub trait Iterator {
...
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// #![feature(iterator_for_each)]
+    ///
+    /// let mut v = vec![];
+    /// (0..5).for_each(|x| v.push(x * 100));
```
Comment (frewsxcv) [inline src/libcore/iter/iterator.rs]: i'm not necessarily opposed to the current example you have written, but i find the current example slightly less idiomatic (subjectively) than something like:

```rust
let v: Vec<_> = (0..5).map(|x| x * 100).collect();
```

Comment (cuviper) [inline src/libcore/iter/iterator.rs]: Sure, I was just aiming for something simple and testable. I would definitely use collect or extend for that in real code. Any ideas for something more meaningful?

Similarly, the added benchmarks are just sums.

Comment (budziq) [inline src/libcore/iter/iterator.rs]: > Any ideas for something more meaningful?

New cookbook contributors usually have problem with consuming functional flow they have built just for the sake of side effects (if they do not wish to obtain any value like in `fold` or `collect`). Switching to imperative `for` just to obtain side effects feels not idiomatic.
 
Some artificial examples that might not be any better :smile_cat:  
``` rust
let (tx, rx) = channel();
(0..5).map(|x| x * 2 + 1).for_each(|x| { tx.send(x).unwrap(); } );
```
```rust
["1", "2", "lol", "baz", "55"]
    .iter()
    .filter_map(|s| s.parse::<u16>().ok())
    .map(|v| v * v)
    .for_each(|v| println!("{}", v));
```

Comment (cuviper) [inline src/libcore/iter/iterator.rs]: @budziq I'm glad to hear of more motivation for this!

Your additional examples are OK, but still not *testing* anything besides successfully compiling.  Note that my first example has an `assert_eq` with the for-loop result, so we actually get *some* sanity check that it really works, as trivial as that is.  Your channel example could read the `rx` side to check the result though.

Comment (budziq) [inline src/libcore/iter/iterator.rs]: @cuviper so something like that might be ok? 
```rust
use std::sync::mpsc::channel;

let (tx, rx) = channel();
(0..5).map(|x| x * 2 + 1).for_each(|x| { tx.send(x).unwrap(); } );
assert_eq!(Ok(vec!(1,3,5,7,9)), (0..5).map( |_| rx.recv()).collect::<Result<Vec<_>,_>>());
```

**Comment by steveklabnik** [comment]
I personally really wanted a `foreach` but yeah, it was rejected, what's changed since then?

**Comment by cuviper** [comment]
@steveklabnik I think at that time the main point was that short iterators are probably better off with a `for` loop and more control-flow options.  It wasn't properly considered for use with longer iterator chains.  When your iterator takes up multiple lines, a `for` loop either makes awkward formatting or requires saving to a local first, both annoying.

Also, the performance benefit of internal iteration via `fold` is pretty compelling, and we can implement that in `for_each` without users having to understand it.

File: src/libcore/iter/iterator.rs (hunk 1 above)
Comment (cuviper) [inline src/libcore/iter/iterator.rs]: That would work.  Do folks like that better?  Or perhaps as one more example?

**Comment by steveklabnik** [comment]
>  It wasn't properly considered for use with longer iterator chains.

I must have done a bad job advocating back then, then. Oh well.

**Comment by alexcrichton** [comment]
Ok sounds like there's no reason to *not* experiment with this at this point, @cuviper if you want to update the example (which I think the comments indicate?) then I'll r+

**Comment by steveklabnik** [comment]
History time!

* https://github.com/rust-lang/rust/pull/14911
* https://github.com/rust-lang/rust/pull/21098
* https://github.com/rust-lang/rust/issues/24566 -> https://github.com/rust-lang/rfcs/issues/1070
* https://github.com/rust-lang/rfcs/pull/582
* https://github.com/rust-lang/rfcs/pull/1064

I would argue that https://github.com/rust-lang/rfcs/pull/1064#issuecomment-93669797 by @nagisa  is a decent summary:

> So far none of the points arguing against the RFC have become false:
>
> * There’s a more general version of this function already available in the standard library as well as various chains that produce the desired behaviour;
> * THere’s the for-loop construct;
>  * itertools package still provides a convenience wrapper for the for-loop.

The first bullet point has changed. The other ones are still true.

**Comment by cuviper** [comment]
Thanks for finding more context!  My search-fu is apparently weak...

I don't think we have to falsify every point against -- only compare them to the points for:

- re `for` loops: [It looks much worse when you're working with a chain of iterators.](https://github.com/rust-lang/rfcs/issues/1070#issuecomment-157837174)
- `for_each` is more succinct for using direct functions, e.g. [`values.for_each(my_log)`](https://github.com/rust-lang/rfcs/issues/1070#issuecomment-157838971)
- internal iteration with `fold` can be much faster, as shown in the benchmarks here.
  - but it's easier for users to write `.for_each(|item| f(item))`
  - rather than writing their own `.fold((), |(), item| f(item))`

**Comment by cuviper** [comment]
In general, I feel like there are a lot of people that do want this, and the people against are just "meh".

Anyway, I updated the first example like @budziq's suggestion.

**Comment by nikomatsakis** [comment]
Obviously I've already been quoted at the top, but I agree with @cuviper that I think the pros outweigh the cons. I think that there is indeed new information since the last time this was discussed:

1. The performance benefits of "internal iteration" were not widely discussed at the time (unless I remember incorrectly; I confess I didn't bother to click all of @steveklabnik's links, I'm just going based on my memory).
2. The fact that, for parallel iteration, `for_each` is **necessary**.

The two together mean that encouraging the use of `for_each` when it is convenient for sequential iteration will make code faster by default, and also facilitate the conversion to parallel execution.  And it's more ergonomic to boot! Seems like a win-win to me.

The main argument against is basically "TMWTDI", which -- from my POV -- is a good enough reason to stop things without compelling advantages, but not an absolute blocker.

I also think one of the points made at the time is at least somewhat false:

> There’s a more general version of this function already available in the standard library as well as various chains that produce the desired behaviour;

I presume that this is referring to `fold` or `all`? I don't consider that a real alternative. It's true that one can model `for_each` this way, but it's misleading and makes the code harder to read -- you have to realize that the function is being abused for something other than its intended purpose, which imo starts to defeat the point of using iterators. Moreover, as a consequence of that, code written in this style cannot be parallelized as efficiently or as well (e.g., the `all` combinator will waste time propagating booleans and checking for shortcircuits; `fold` isn't even available).

File: src/libcore/iter/iterator.rs [hunk 2]
```diff
@@ -482,6 +482,53 @@ pub trait Iterator {
...
+    /// Basic usage:
+    ///
+    /// ```
+    /// #![feature(iterator_for_each)]
+    ///
+    /// use std::sync::mpsc::channel;
+    ///
+    /// let (tx, rx) = channel();
+    /// (0..5).map(|x| x * 2 + 1)
+    ///       .for_each(move |x| tx.send(x).unwrap());
```
Comment (nikomatsakis) [inline src/libcore/iter/iterator.rs]: Is the `move` necessary here? I would not *expect* so.

Comment (cuviper) [inline src/libcore/iter/iterator.rs]: Maybe it's too sneaky, but that lets `tx` drop automatically, and then `rx` won't block.

**Comment by cuviper** [comment]
Yay!

Now, since I left it unstable with `issue = "0"`, do we need to open an issue and update that with a PR?

**Comment by Mark-Simulacrum** [comment]
Oh, yes, please do. We probably shouldn't have merged yet actually, but no worries!

**Comment by cuviper** [comment]
See #42987 for that update.

**Comment by phimuemue** [comment]
Hi, I stumbled upon this, but I wondered why `for_each` does not support break. I hope I don't miss something essential, but I imagined it would be no problem that `for_each` takes a function returning a value that determines whether we would like to break or not.

This value could be - as it is now - a `()` (never causing the loop to break) or a `bool` indicating whether we want to break or an `Option<T>` indicatin whether we want to "break with a certain item".

In code, this could be captured by a trait `BreakIndicator` as follows:

```
trait BreakIndicator {
    fn is_break(&self) -> bool;
    fn final_continue() -> Self;
}

impl BreakIndicator for () {
    // always continue
    fn is_break(&self) -> bool { false }
    fn final_continue() -> () {}
}

impl BreakIndicator for bool {
    // true means break; false means continue
    fn is_break(&self) -> bool { *self }
    fn final_continue() -> bool { false }
}

impl<T> BreakIndicator for Option<T> {
    // Some(v) means "break wich value v"; None means continue
    fn is_break(&self) -> bool { self.is_some() }
    fn final_continue() -> Option<T> { None }
}
```

I implemented a `fn breaking_for_each` on top of `Iterator` to see how that would work out:

```
trait IteratorWithBreakingForEach : Iterator {
    fn breaking_for_each<F, BI>(self, f: F) -> BI
        where F: FnMut(Self::Item) -> BI,
              BI: BreakIndicator,
    ;
}

impl<I> IteratorWithBreakingForEach for I where I: Iterator {
    fn breaking_for_each<F, BI>(self, mut f: F) -> BI
        where F: FnMut(Self::Item) -> BI,
              BI: BreakIndicator,
    {
        for item in self {
            let break_indicator = f(item);
            if break_indicator.is_break() {
                return break_indicator;
            }
        }
        BI::final_continue()
    }
}
```

Usage could be as follows:

```
fn main() {
    (1..10).breaking_for_each(|i| {
        println!("{}", i) // does not break at all
    });
    (1..10).breaking_for_each(|i| {
        println!("{}", i);
        i>=5 // break if i>=5
    });
    let x = (1..10).breaking_for_each(|i| {
        println!("{}", i);
        if i>=5 {
            Some(i) // break with value
        } else {
            None // continue
        }
    });
    println!("{:?}", x);
}
```

Has this ever been thought about? And if so, why was it apparently rejected?

**Comment by cuviper** [comment]
@phimuemue There's a form of that built around the `Try` trait with [`try_for_each`](https://github.com/rust-lang/rust/pull/48157) [(docs)](https://doc.rust-lang.org/nightly/std/iter/trait.Iterator.html#method.try_for_each).  There's also some discussion in https://github.com/rust-lang/rust/issues/42327#issuecomment-376772143 whether `Try` should be re-framed more like `Break`/`Continue`.

**Comment by scottmcm** [comment]
@phimuemue As an example of "break with a certain item", check out how `find` is implemented:

https://github.com/rust-lang/rust/blob/ab8b961677ac5c74762dcea955aa0ff4d7fe4915/src/libcore/iter/iterator.rs#L1738-L1746

(That `LoopState` type is currently internal, but personally I'd like `Try` to use it, as @cuviper said.)

If you wanted to do the same in your code today†, it can be done with `Result` like this:

```rust
    self.try_for_each(move |x| { 
        if predicate(&x) { Err(x) } 
        else { Ok(()) } 
    }).err() 
```

† Well, once someone makes a stabilization PR for `try_for_each`...

**Comment by scottmcm** [comment]
Hello 4 years later!  In case anyone else ends up here, the `LoopState` type mentioned in my previous comment is now available on stable as https://doc.rust-lang.org/stable/std/ops/enum.ControlFlow.html

**Pull request merged**